server/sonar-server/src/main/java/org/sonar/server/authentication/UserSessionInitializer.java

```diff
@@ -0,0 +1,128 @@
+import static org.sonar.api.CoreProperties.CORE_FORCE_AUTHENTICATION_PROPERTY;
+import static org.sonar.api.web.ServletFilter.UrlPattern.Builder.staticResourcePatterns;
+import static org.sonar.server.authentication.ws.LoginAction.AUTH_LOGIN_URL;
+import static org.sonar.server.authentication.ws.ValidateAction.AUTH_VALIDATE_URL;
+
+import com.google.common.collect.ImmutableSet;
+import java.util.Optional;
+import java.util.Set;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.sonar.api.config.Settings;
+import org.sonar.api.server.ServerSide;
+import org.sonar.api.web.ServletFilter;
+import org.sonar.db.DbClient;
+import org.sonar.db.user.UserDto;
+import org.sonar.server.exceptions.UnauthorizedException;
+import org.sonar.server.user.ServerUserSession;
+import org.sonar.server.user.ThreadLocalUserSession;
+
+@ServerSide
+public class UserSessionInitializer {
+
+  // SONAR-6546 these urls should be get from WebService
+  private static final Set<String> SKIPPED_URLS = ImmutableSet.of(
+    "/batch/index", "/batch/file", "/batch_bootstrap/index",
+    "/maintenance/*",
+    "/setup/*",
+    "/sessions/*",
+    "/api/system/db_migration_status", "/api/system/status", "/api/system/migrate_db",
+    "/api/server/*",
+    AUTH_LOGIN_URL, AUTH_VALIDATE_URL
+  );
+
+  private static final ServletFilter.UrlPattern URL_PATTERN = ServletFilter.UrlPattern.builder()
+    .includes("/*")
+    .excludes(staticResourcePatterns())
+    .excludes(SKIPPED_URLS)
+    .build();
+
+  private final DbClient dbClient;
+  private final Settings settings;
+  private final JwtHttpHandler jwtHttpHandler;
+  private final BasicAuthenticator basicAuthenticator;
+  private final ThreadLocalUserSession userSession;
+
+  public UserSessionInitializer(DbClient dbClient, Settings settings, JwtHttpHandler jwtHttpHandler, BasicAuthenticator basicAuthenticator,
+    ThreadLocalUserSession userSession) {
+    this.dbClient = dbClient;
+    this.settings = settings;
+    this.jwtHttpHandler = jwtHttpHandler;
+    this.basicAuthenticator = basicAuthenticator;
+    this.userSession = userSession;
+  }
+
+  public boolean initUserSession(HttpServletRequest request, HttpServletResponse response) {
+    String path = request.getRequestURI().replaceFirst(request.getContextPath(), "");
+    try {
+      if (!URL_PATTERN.matches(path) || isDeprecatedBatchWs(path)) {
+        return true;
+      }
+
+      Optional<UserDto> user = authenticate(request, response);
+      if (user.isPresent()) {
+        userSession.set(ServerUserSession.createForUser(dbClient, user.get()));
+      } else {
+        if (settings.getBoolean(CORE_FORCE_AUTHENTICATION_PROPERTY)) {
+          throw new UnauthorizedException("User must be authenticated");
+        }
+        userSession.set(ServerUserSession.createForAnonymous(dbClient));
+      }
+      return true;
+    } catch (UnauthorizedException e) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1166)